### PR TITLE
Stabilize the `#[diagnostic]` namespace and `#[diagnostic::on_unimplemented]` attribute

### LINKED ISSUE #111996: Tracking Issue for `#![feature(diagnostic_namespace)]`
Feature gate: `#![feature(diagnostic_namespace)]`

This is a tracking issue for [RFC 3368](https://github.com/rust-lang/rfcs/pull/3368), which introduces a `#[diagnostic]` attribute namespace for attributes that affect the compiler diagnostics. Using an unknown attribute in this attribute namespace should only result in a warning.

### Public API

```rust

#[diagnostic::something] // generates a lint warning.
pub struct Magic;
```

### Steps / History

- [x] Implementation: https://github.com/rust-lang/rust/pull/111780
- [x] Final comment period (FCP)[^1]: https://github.com/rust-lang/rust/pull/119888#issuecomment-1944179957
- [ ] Stabilization PR: https://github.com/rust-lang/rust/pull/119888

### Unresolved Questions

- None yet. (All unresolved questions in the RFC are related to the design of the `#[diagnostic::on_unimplemented]` attribute, which likely will use a different feature)

[^1]: https://std-dev-guide.rust-lang.org/feature-lifecycle/stabilization.html

### PR DESCRIPTION
**Pull request by weiznich**
This PR stabilizes the `#[diagnostic]` attribute namespace and a minimal option of the `#[diagnostic::on_unimplemented]` attribute.

The `#[diagnostic]` attribute namespace is meant to provide a home for attributes that allow users to influence error messages emitted by the compiler. The compiler is not guaranteed to use any of this hints, however it should accept any (non-)existing attribute in this namespace and potentially emit lint-warnings for unused attributes and options. This is meant to allow discarding certain attributes/options in the future to allow fundamental changes to the compiler without the need to keep then non-meaningful options working.

The `#[diagnostic::on_unimplemented]` attribute is allowed to appear on a trait definition. This allows crate authors to hint the compiler to emit a specific error message if a certain trait is not implemented. For the `#[diagnostic::on_unimplemented]` attribute the following options are implemented:

* `message` which provides the text for the top level error message
* `label` which provides the text for the label shown inline in the broken code in the error message
* `note` which provides additional notes.

The `note` option can appear several times, which results in several note messages being emitted. If any of the other options appears several times the first occurrence of the relevant option specifies the actually used value. Any other occurrence generates an lint warning. For any other non-existing option a lint-warning is generated.

All three options accept a text as argument. This text is allowed to contain format parameters referring to generic argument or `Self` by name via the `{Self}` or `{NameOfGenericArgument}` syntax. For any non-existing argument a lint warning is generated.

This allows to have a trait definition like:

```rust
#[diagnostic::on_unimplemented(
    message = "My Message for `ImportantTrait<{A}>` is not implemented for `{Self}`",
    label = "My Label",
    note = "Note 1",
    note = "Note 2"
)]
trait ImportantTrait<A> {}

```

which then generates for the following code 

```rust
fn use_my_trait(_: impl ImportantTrait<i32>) {}

fn main() {
    use_my_trait(String::new());
}
```

this error message:

```
error[E0277]: My Message for `ImportantTrait<i32>` is not implemented for `String`
  --> src/main.rs:14:18
   |
14 |     use_my_trait(String::new());
   |     ------------ ^^^^^^^^^^^^^ My Label
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `ImportantTrait<i32>` is not implemented for `String`
   = note: Note 1
   = note: Note 2
```

[Playground with the unstable feature](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=05133acce8e1d163d481e97631f17536)

Fixes #111996

### CODE REVIEW

**Comment by ehuss** [comment]
Before stabilization, can we make sure a reference PR is opened and reviewed?

**Comment by compiler-errors** [comment]
> this error message:

This is missing

**Comment by weiznich** [comment]
> Before stabilization, can we make sure a reference PR is opened and reviewed?

Do you have any suggestions where exactly the relevant documentation should be placed in the reference?

**Comment by ehuss** [comment]
There are a few places that need to be updated:

* Add it to the list of [built-in attributes](https://github.com/rust-lang/reference/blob/master/src/attributes.md#built-in-attributes-index) under the diagnostic section.
* Create a section in the [diagnostics chapter](https://github.com/rust-lang/reference/blob/master/src/attributes/diagnostics.md) that details the syntax and behavior of the attribute. There are templates for the format that attributes use (see deprecated and must_use on that page for examples).

Is this implemented as a [Tool attribute](https://doc.rust-lang.org/nightly/reference/attributes.html#tool-attributes)? If so, that section should probably be updated. If not, then it might be OK in the diagnostics chapter to have a ``The `diagnostic` namespace`` section and then a subsection under that ``The `diagnostic::on_unimplemented` attribute``.

Thanks!

**Comment by weiznich** [comment]
Thanks for the pointers :+1:

I've opened a PR here: https://github.com/rust-lang/reference/pull/1449

File: library/core/src/future/future.rs [hunk 1]
```diff
@@ -28,10 +28,13 @@ use crate::task::{Context, Poll};
 #[must_use = "futures do nothing unless you `.await` or poll them"]
 #[stable(feature = "futures_api", since = "1.36.0")]
 #[lang = "future_trait"]
-#[diagnostic::on_unimplemented(
-    label = "`{Self}` is not a future",
-    message = "`{Self}` is not a future",
-    note = "{Self} must be a future or must implement `IntoFuture` to be awaited"
+#[cfg_attr(
```
Comment (Nilstrieb) [inline library/core/src/future/future.rs:31]: Why did you `cfg_attr` it here? It doesn't appear to have changed. You should `cfg_attr` the `feature()` declaration instead.

Comment (weiznich) [reply library/core/src/future/future.rs]: I just did not occur to me that that would also work :see_no_evil: 
I've pushed a new version that reverts the changes and uses that approach instead.

**Comment by weiznich** [comment]
@rustbot label +T-lang

(As this was decided in the discussion around the RFC https://github.com/rust-lang/rfcs/pull/3368)

**Comment by nikomatsakis** [comment]
@rustbot labels -T-compiler -T-libs

We discussed this in the @rust-lang/lang meeting and decided to remove compiler/libs for purposes of FCP, since this seemed analogous to other language features (those teams should feel free to add an FCP of their own if they feel it appropriate/necessary).

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I propose we merge this idea! I can't wait to see libraries take control over their own diagnostics.

**Comment by tmandry** [comment]
One point of feedback: It's interesting that there seems to be a typo in the example message ("implemented" versus "not implemented"). I know that's just an example, but it raises an interesting point that we might want the compiler to have a "default message", note, or something else that helps clarify exactly what the problem is.

In any case I'm in favor, so

@rfcbot reviewed

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this in the lang triage meeting today.  We were in favor of doing this.  This is now in FCP.

Thanks to @weiznich for pushing this forward.

**Comment by weiznich** [comment]
@tmandry Thanks for pointing out the typo in the example message. I've updated the example to be hopefully more clear now.

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r? compiler-errors @bors r+

**Comment by ehuss** [comment]
Is it intentional that the format parameters support the general syntax of std's format strings? There seems to be a variety of situations that don't seem to be handled like I would expect:

* Strings such as `{}` or `{1:}` generate an error instead of a warning. Doesn't that go against the intention that the diagnostic attributes are flexible in what they accept to support forwards compatibility?
* Strings such as `{A:1}` *don't* generate a warning or an error, even though it includes a nonsense width argument. This is true for all format spec things (precision, `?` debug, etc.)
* Format specifiers allow flexible whitespace, like `{A   }`.
* Braces are escaped by doubling them with `{{` or `}}`.
* Incorrect braces seem to be ignored, and simply stop parsing, and are ignored (no warning). For example "foo }"

Is it intended that these format parameters are simple string replacements (like `{Self}` gets replaced with the type name) or is it intended that it uses the full formatting machinery of std's format strings (allowing things like width specifiers)? In either case, it seems like there are several situations that are not handled as I would expect. Additionally, none of this was documented.

**Comment by weiznich** [comment]
@ehuss I'm not sure I can follow you here. Do you mean that this was not mentioned at all? (In that case, both the Reference PR and the description explicitly mentions it). Or do you mean that it contains more features as described there? Or that these 4 specific cases result in different results than you personally would expect? For the later case I would argue that this is merely a bug that can be fixed independently from stabilization. Given a proper report with an explicit list of test cases I can try working on that.

**Comment by ehuss** [comment]
> Do you mean that this was not mentioned at all? (In that case, both the Reference PR and the description explicitly mentions it). 

The description says "This text is allowed to contain format parameters referring to ..." but doesn't define what a "format parameter" is other than `{Self}` or `{NameOfGenericArgument}`. Unless I missed something, I don't see where it says that the entire string is a [`std::fmt`](https://doc.rust-lang.org/std/fmt/index.html) style format string. If the string is to be interpreted using some syntax system, then that needs to be explicitly stated and documented.

> Given a proper report with an explicit list of test cases I can try working on that.

Sure, I'll file an issue.

**Comment by ehuss** [comment]
To put my concerns in a different way, this now requires that compiler implementers be able to interpret std format strings at runtime, which is not a requirement I think has been made until now. It is very unlikely to be an issue, but seems like a significant requirement that was not stated.

**Pull request merged**